### PR TITLE
Fix perf when populate on huge qs

### LINKED ISSUE #86: Performance issue when populate millions documents
##### Problem
When we need to put a lot of documents in index, we need to use `queryset_pagination` meta option to paginate. Django pagination need a sorted queryset with `order_by` (cf doc)[https://docs.djangoproject.com/en/2.0/topics/pagination/#required-arguments] otherwise same pk can be present more than once.

Put `order_by` on queryset will make django paginator call order_by for each page. Call `order_by` on huge queryset (like 10 millions) will lead to a huge perfomance issue.

##### Temporary solution:

We can override `_get_actions` method (from `django_elasticsearch_dsl.documents.DocType`) to not use django paginator when a queryset is passed. More over because of the way a database index work, we should first fetch only pks, and then do sub request based on it.

```python
from django.db.models.query import QuerySet


def _get_actions(self, object_list, action):
    if self._doc_type.queryset_pagination and isinstance(object_list, QuerySet):
        pks = object_list.order_by('pk').values_list('pk', flat=True)
        for start_pk_index in range(0, len(pks), self._doc_type.queryset_pagination + 1):
            end_pk_index = start_pk_index + self._doc_type.queryset_pagination
            ranged_qs = object_list.filter(pk__range=[
                pks[start_pk_index],
                pks[end_pk_index]
            ])
            for object_instance in ranged_qs:
                yield self._prepare_action(object_instance, action)
    else:
        super()._get_actions(object_list, action)
```

Available to make the PR if needed.

### PR DESCRIPTION
**Pull request by ahivert**
Fix #86

- create a custom paginator based on pks
- Add test and refactor tests_documents to use models.py and documents.py
- make travis run flake8

### CODE REVIEW

**Comment by safwanrahman** [comment]
What about using `iterator()` to optimize it?

**Comment by sabricot** [comment]
https://docs.djangoproject.com/en/2.1/ref/models/querysets/#django.db.models.query.QuerySet.iterator use server-side cursors with postgresql but not with mysql so i guess it wil not optimize every DB backend.
Plus with `iterator` we lose the `prefetch_related()` queries and in some scenarios they can be used.

File: django_elasticsearch_dsl/documents.py [hunk 1]
```diff
@@ -202,13 +202,26 @@ def _prepare_action(self, object_instance, action):
...
-            for page in paginator.page_range:
-                for object_instance in paginator.page(page).object_list:
-                    yield self._prepare_action(object_instance, action)
+        qs_pagination = self._doc_type.queryset_pagination
+        if qs_pagination is not None:
+            if isinstance(object_list, models.QuerySet):
+                # Django paginator can lead to performance issue with huge
+                # queryset, make custom one based on pks presents in db right
+                # now
+                pks = list(object_list.values_list('pk', flat=True))
```
Comment (sabricot) [inline django_elasticsearch_dsl/documents.py]: I can't really test it. Do you have tested to retrieve millions of pks in this list ?
And don't you think that this `object_list.values_list` queryset should be ordered by pk ?

File: .travis.yml [hunk 2]
```diff
@@ -96,7 +96,9 @@ before_script:
   - sleep 10
 
 # command to run tests using coverage, e.g. python setup.py test
-script: tox -e $TOX_ENV -- --elasticsearch
+script:
+  - flake8
```
Comment (sabricot) [inline .travis.yml]: 👍

File: tests/test_documents.py [hunk 3]
```diff
@@ -11,46 +11,27 @@
                                                  RedeclaredFieldError)
 from tests import ES_MAJOR_VERSION
 
-
-class Car(models.Model):
```
Comment (sabricot) [inline tests/test_documents.py]: 👍

File: django_elasticsearch_dsl/documents.py (hunk 1 above)
Comment (ahivert) [reply django_elasticsearch_dsl/documents.py]: Of course ! improve this code with the new commit (--amend)

**Comment by ahivert** [comment]
That should be better now :)

**Pull request closed without merging**